src/Orleans/Messaging/GatewayConnection.cs

```diff
@@ -248,6 +248,7 @@ protected override bool GetSendingSocket(Message msg, out Socket socketCapture,
 
         protected override void OnGetSendingSocketFailure(Message msg, string error)
         {
+            msg.TargetSilo = null; // clear previous destination!
```

Comment:
That line was missing and that was the bug.
Basically, a message was already addressed to a GW that was already shutting down. Thus, it could not be sent. Upon send socket error the code attempts to resend it, but it used the same old TargetSilo address in the header, which caused this message to be attempted to be send to the same GW again, instead of to a new GW.
So just cleared an old TargetSilo  which would cause this message to be re-addressed again to a new live GW.